### DISCUSSION TITLE
How do I create Python bindings that work with wrapped RDKit molecules?

### DISCUSSION THREAD

**Discussion by AlanKerstjens** [rdkit/rdkit#6092] [Q&A] [answered]
I've written some C++ code to work with RDKit molecules and would like to create Python bindings using Boost.Python.

Suppose I have the following (compiling) code:

```cpp
#include <boost/python.hpp>
#include <GraphMol/RWMol.h>

std::size_t CountAtoms(const RDKit::ROMol& molecule) {
  return molecule.getNumAtoms(); 
};

void EditMolecule(RDKit::RWMol& molecule) {
  molecule.addAtom();
};

BOOST_PYTHON_MODULE(MyModule) {
  boost::python::def("CountAtoms", CountAtoms);
  boost::python::def("EditMolecule", EditMolecule);
};
```

Then, in Python I would like to do something like this:

```python
from rdkit import Chem
import MyModule

romol = Chem.MolFromSmiles("C")
rwmol = Chem.RWMol(romol)

MyModule.CountAtoms(romol) # ERROR
MyModule.EditMolecule(rwmol) # ERROR
```

The calls to the functions I exposed fail due to type errors:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
Boost.Python.ArgumentError: Python argument types in
    MyModule.CountAtoms(Mol)
did not match C++ signature:
    CountAtoms(RDKit::ROMol)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
Boost.Python.ArgumentError: Python argument types in
    MyModule.EditMolecule(RWMol)
did not match C++ signature:
    EditMolecule(RDKit::RWMol {lvalue})
```

I suppose I need to write my functions to work with wrapped RDKit molecules (the ones that are actually exposed in Python), which if I'm not mistaken are the `Mol` and `ReadWriteMol` classes. 

* Where can I find the source code for the `Mol` class? I found the code for `ReadWriteMol` [here](https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/Wrap/Mol.cpp) but the code for the `Mol` class eludes me.
* Is there anything I should be aware of when using these classes?
* Do I need to rewrite all of my code to work `Mol` and `ReadWriteMol` instead of `ROMol` and `RWMol`?
* What's the reason for this extra layer of indirection? Looking at the code for `ReadWriteMol` it's just calling the corresponding `RWMol` functions. Why wasn't `RWMol` exposed directly?

**Comment by bp-kelley** [comment]
It's not for the faint of heart for sure, but if get the rdkit repository and look in the External directory, you'll see examples of wrapping external libraries using RDKit molecules.  A decent starting point is here that wraps FreeSASA into rdkit.

https://github.com/rdkit/rdkit/tree/master/External/FreeSASA/

The actually wrapping in boost python is here:

https://github.com/rdkit/rdkit/tree/master/External/FreeSASA/Wrap

We don't have an example that allows wrapping from an rdkit install though, this is done at rdkit build time.

**Reply by AlanKerstjens** [reply]
Thanks for the pointer. I've copied their example and removed everything SASA related. Here is what I have, but it still doesn't work:

```cpp
#include <RDBoost/python.h>
#include <RDGeneral/Exceptions.h>
#include <GraphMol/RDKitBase.h>
#include <RDBoost/Wrap.h>

namespace python = boost::python;

namespace RDKit {

namespace {
  int CountAtoms(const RDKit::ROMol &mol) {
    return mol.getNumAtoms();
  };
}; // namespace

struct atom_count_wrapper {
  static void wrap() {
    python::def(
        "CountAtoms", CountAtoms,
        (python::arg("mol")));
  };
};

};  // namespace RDKit

BOOST_PYTHON_MODULE(MyModule) {
  RDKit::atom_count_wrapper::wrap();
};
```

> We don't have an example that allows wrapping from an rdkit install though, this is done at rdkit build time.

This worries me. Does that mean it isn't possible to wrap my code without rebuilding the entire RDKit? Ideally I would like to compile just my own code and have it interface with the already built RDKit.

**Comment by bp-kelley** [comment]
Here is the wrapping code itself:

https://github.com/rdkit/rdkit/blob/master/External/FreeSASA/Wrap/rdFreeSASA.cpp

**Comment by DavidACosgrove** [comment]
Without the compiler messages, it's hard to say for sure, but in this case I think you need to pass the address of CountAtoms into the python::def:
```
struct atom_count_wrapper {
  static void wrap() {
    python::def(
        "CountAtoms", &CountAtoms,
        (python::arg("mol")));
  };
};
```
Whilst I'm sure @bp-kelley is correct about there being no examples of exactly what you're doing, there are lots of examples of wrapping C++ functions in boost::python in the repo which may be helpful.  Look for directories called Wrap underneath the directories in $RDBASE/Code/GraphMol.  There are examples of how to pass larger C++ entities like lists and vectors of vectors and such, for example, which may help.  https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/MolDraw2D/Wrap/rdMolDraw2D.cpp is particularly rich with them.

**Reply by bp-kelley** [reply]
@DavidACosgrove  while we do have examples of building extensions during the rdkit build, we don't have ones for building against an installed rdkit package.  This is what I was trying, and failing, to describe.

**Comment by AlanKerstjens** [comment, marked as answer]
Success! The main problem was that I was using a more modern version of Python / Boost.Python than the one I had used to compile the RDKit. The code compiled fine with no errors, but the (old) wrapped objects weren't recognized as the arguments for the (modern) functions. Recompiling the RDKit with the same versions fixed it. 

It seems the code for working with `ROMol`s was correct. However, to make it work with `RWMol`s I had to cast the `RWMol` to a `ROMol`. This is how functions such as `MolOps::Kekulize` are exposed to Python.

https://github.com/rdkit/rdkit/blob/a49380e1efb6a20bdab3b8e86deb64a25d1156e5/Code/GraphMol/Wrap/MolOps.cpp#L420-L423

Here is the code that wound up working:

```cpp
#include <GraphMol/RWMol.h>
#include <boost/python.hpp>

int CountAtoms(const RDKit::ROMol& mol) {
  return mol.getNumAtoms();
};

void EditMolecule(RDKit::ROMol& molecule) {
  RDKit::RWMol& rwmol = static_cast<RDKit::RWMol&>(molecule);
  rwmol.addAtom();
};

BOOST_PYTHON_MODULE(MyModule) {
  boost::python::def("CountAtoms", CountAtoms);
  boost::python::def("EditMolecule", EditMolecule);
};
```
```python
from rdkit import Chem
import MyModule

mol = Chem.MolFromSmiles("C")
print(MyModule.CountAtoms(mol)) # 1
MyModule.EditMolecule(mol)
print(Chem.MolToSmiles(mol)) # *.C
```

However, this all seems a bit unnatural to me. I thought the whole point of having separate `ROMol` and `RWMol` classes was to prevent mutation. I can construct a wrapped `RWMol` in Python. Why can't I pass it to exposed C++ functions expecting a `RWMol`?

```cpp
#include <GraphMol/RWMol.h>
#include <boost/python.hpp>

void EditROMol(RDKit::ROMol& romol) {
  RDKit::RWMol& rwmol = static_cast<RDKit::RWMol&>(romol);
  rwmol.addAtom();
};

void EditRWMol(RDKit::RWMol& rwmol) {
  rwmol.addAtom();
};

BOOST_PYTHON_MODULE(MyModule) {
  boost::python::def("EditROMol", EditROMol);
  boost::python::def("EditRWMol", EditRWMol);
};
```
```python
from rdkit import Chem
import MyModule

romol = Chem.MolFromSmiles("C")
rwmol = Chem.RWMol(romol)
MyModule.EditROMol(romol) # Works
MyModule.EditROMol(rwmol) # Works
MyModule.EditRWMol(romol) # Error (expected)
MyModule.EditRWMol(rwmol) # Error (unexpected)
```

**Reply by bp-kelley** [reply]
@AlanKerstjens this is a historical artifact from the original wrappers.  Originally, RWMol was not exposed to python, just Mol.  In the C++ land all ROMols are RWMols so the boost::python interface gleefully and helpfully did the cast for you whether you wanted it to or not.

When we added RWMol to the boost wrappers, we initially tried to separate the two like you currently want, but it broke a LOT of code so we just decided against it, in fact it turned everything into RWMols on the python side.

So now, you end up with this odd system that in python land, RWMol and Mol(ROMol) are different but as soon as you pass them through boost, all bets are off.

rdkit 3000 may fix this :)
